### PR TITLE
Merge JUnit unit tests report

### PR DESCRIPTION
**Pull request by wzieba**
Wait with merge for this PR:
- https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/103

### Description

This PR adds behavior of merging reports of unit tests into a single file before requesting Buildkite annotations or copying them for Buildkite Test Collector.

This way, these operations is far more performant. See **Impact**

#### Buildkite annotations

| | `WordPress` | 
| -- | -- |
| Before (2069f5709555311539c94b1a3bb56670b954e072) | 4m33s |
| After | 1s |
| Diff | -99.6% 😅 |

#### Collecting artifacts

| | `WordPress` | 
| -- | -- |
| Before (2069f5709555311539c94b1a3bb56670b954e072) | 8m49s |
| After | 1s |
| Diff | -99.8% 😅 |

### Testing

Verify if the number of tests reported by Buildkite test collector (`View in Test Analytics` button on Buildkite) is the same before and after this PR.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: .buildkite/commands/run-unit-tests.sh
Comment (AliSoftware) [inline .buildkite/commands/run-unit-tests.sh:39]: As a side note and only for my curiosity's sake, I'd be interested to better understand why there's such a huge difference in practice before/after. In particular, I'm curious why it took so long — ≈4mn to do buildkite annotations and ≈9mn to collect artifacts — before the merging and merging them together reduces the time taken by so much.

 - Not only am I interested by this for curiosity's sake to better understand what caused this extra time and potentially act on it (would the root cause suggest that we should maybe optimize our scripts like `annotate_test_failures` in the first place?)
 - But also, this change making those steps now taking only 1s compared to mulitiple minutes previously… leaves me a bit wary and wondering if it's doing anything at all… as opposed to that new 1s time being because those scripts exit early without doing anything (e.g. because we overlooked something in this diff and pass them the wrong path to parse, or the merged file wasn't created as expected for some reason, …)

Comment (wzieba) [reply .buildkite/commands/run-unit-tests.sh:39]: Please feel free to examine the `trunk` commit that this branch is branched from `Unit Test WordPress` job: https://buildkite.com/automattic/wordpress-android/builds/19426#01910379-29a1-4bfe-b44e-f0c6276aad00

> would the root cause suggest that we should maybe optimize our scripts like annotate_test_failures in the first place

I don't see there's anything wrong with `annotate_test_failures`, but I don't think it was ever considered to be used in a loop with 528 iterations (each `annotate_test_failures` AFAIU performs communication with Buildkite API).

> But also, this change making those steps now taking only 1s compared to mulitiple minutes previously… leaves me a bit wary and wondering if it's doing anything at all…

It does - please see `View in Test Analytics` - the number of tests is the same. `Artifacts` contains merged report of all tests, this also works fine.

For annotations part, I'll prepare a temporary unit test failure to make sure that annotations still work.

Comment (wzieba) [reply .buildkite/commands/run-unit-tests.sh:39]: Worth to note the `pre-exit` hook that throws a number of errors on `trunk`: https://buildkite.com/automattic/wordpress-android/builds/19426#01910379-29a1-4bfe-b44e-f0c6276aad00/3521-3522 and last for several minutes.

Comment (AliSoftware) [reply .buildkite/commands/run-unit-tests.sh:39]: > I don't see there's anything wrong with annotate_test_failures, but I don't think it was ever considered to be used in a loop with 528 iterations (each annotate_test_failures AFAIU performs communication with Buildkite API).

Ooof, 528 iterations each with a Buildkite API call seems a lot, that could indeed explain it all 😅  Thanks for the links 👍

Comment (wzieba) [reply .buildkite/commands/run-unit-tests.sh:39]: > For annotations part, I'll prepare a temporary unit test failure to make sure that annotations still work.

Works as expected! https://buildkite.com/automattic/wordpress-android/builds/19511#019122ab-dbf7-4018-aced-830561a73c21

image

**Review by AliSoftware — APPROVED** [review]
LGTM, approving to unblock… provided of course that you remember to update the `export CI_TOOLKIT` in `.buildkite/shared-pipeline-vars` after https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/103 lands before merging this one 🙂

File: .buildkite/commands/run-unit-tests.sh
Comment (AliSoftware) [inline .buildkite/commands/run-unit-tests.sh]:
```suggestion
merge_junit_reports -d ${test_log_dir%/*} -o $results_file
```

**Pull request merged**
